### ISSUE TITLE
Mirroring images for 194 landmark with Helen

### ISSUE THREAD

**Issue by edumucelli** [davisking/dlib#1003]
First of all thank you for this amazing library, I am having a lot of fun using it. I am training a 194-landmarks shape predictor using Helen dataset. As the results are not quite good yet, although I am using Kazemi's parameters, which AFAIK are (please correct me I am wrong)

```
options.cascade_depth = 10
options.tree_depth = 5
options.num_trees_per_cascade_level = 500
options.nu = 0.1
options.oversampling_amount = 20
options.feature_pool_size = 400
options.lambda_param = 0.1
options.num_test_splits = 20
options.feature_pool_region_padding = 0
options.be_verbose = True
```

I am looking for a way to increase the dataset instead of further tweaking the parameters. As Kazemi have used "778  LFPW" extra images (besides Helen) which I could not find the source, I will mirror the images in Helen (horizontally flipped), basically as you did for the IBUG to train the 68 landmarks predictor. I am wondering how to position the `x` and `y` attributes of the `part` attribute on the mirrored version in the Pascal VOC format? Have you calculated it based on a offset to the `left` attribute of the face's mirrored bounding box? If so, could you clarify how?

Thank you in advance.

**Comment by davisking** [comment]
Thanks, glad you like dlib :)

I mirrored the dataset by manually creating a mapping between mirrored and
non-mirrored points.  Like, left part of left eye becomes right part of
right eye.  You can do this because all the points have matching symmetric
points on the other side of the face.  There is no general way to do this,
you have to just look at the points and write down the mapping for the
annotation format you are using.  Each annotation format will have some
different mapping.  Right, you have to know the corresponding symmetric
point pairs for your annotation style (e.g. left corner of left eye goes
with right corner of right eye).

**Comment by davisking** [comment]
Actually, upon further reflection, it's probably possible to compute this
transformation reliably from just the raw training data using something
like the iterative closest point algorithm in the right way.  But really,
you should just do it manually :)

**Comment by edumucelli** [comment]
Hi again @davisking, I am still working on the 194-landmark prediction since the first time we have discussed and I have some progress and having much fun with dlib! The predictor works almost OK, but in some instances it performances well for one image but poorly for its mirrored version. Consider here that the mirrored version is correct in terms of coordinates since it comes from the default Helen dataset. The problem happens mostly in the jaw. For instance, here are some of the examples

![g](https://user-images.githubusercontent.com/959623/38465287-5aba39f4-3b1a-11e8-9a41-e6b7cf94e4e0.png)
![h](https://user-images.githubusercontent.com/959623/38465286-5aa1fb64-3b1a-11e8-932c-3281767f99a6.png)

![b](https://user-images.githubusercontent.com/959623/38465300-8151cf28-3b1a-11e8-8a45-6557c50559b7.png)
![a](https://user-images.githubusercontent.com/959623/38465302-8682bdfe-3b1a-11e8-86b1-6dc24adcef52.png)

Given your experience, I am wondering if you have an idea of why that could happen, good for one instance and bad for its mirrored counterpart and why mostly in the jaw (maybe it is the hardest part as it has a long frontier with the background)?

Besides, I am thinking about adding to the train set the images for which the predictor correctly labeled  (which I manually verify that the points are in the good position). Do you think that would be OK, or could have any side effect of inject its own good predictions on the train set. I was not able to get the remaining images from the paper, so I am looking for alternatives to increase the dataset. Thanks in advance!

**Comment by jankaWIS** [comment]
Hi @davisking and thanks for those answers. I have been also looking into mirroring images and their labels to get a bigger dataset. I have read through your [post](http://blog.dlib.net/2018/01/correctly-mirroring-datasets.html) which you mentioned above and through your comments here and got a bit confused. In the post you write:
>the --flip command now performs automatic source label matching using a 2D point registration algorithm. That is, it left-right flips the dataset and annotations. Then it registers the mirrored landmarks with the original landmarks and transfers labels appropriately. In fact, the "source label matching" image on the right was created by the new version of `imglab`.

I didn't find any info about that in the [README](https://github.com/davisking/dlib/tree/master/tools/imglab) but I understood that you supply images and then you can go through them, adjust boxes and create the xml file. Then it would also take care of the mirroring. But what if I already have the xml file and now I'd just want to double it by mirroring the images (ie creating the labels from the xml) and flipping the images (that I can take care of in another way but maybe it can be done in one shot)? Is there a function or a command in `dlib`? 

Also, I have not tested it yet but does `dlib` run on windows 10?

Thanks.

**Comment by SebhasSilva** [comment]
Hola, alguien sabe donde puedo descargar el shape predictor de 194
